### PR TITLE
On 386 learner data access

### PR DESCRIPTION
**Pull request by arshedmust**

### CODE REVIEW

File: src/SFA.DAS.AssessorService.Data/IlrRepository.cs [hunk 1]
```diff
@@ -52,5 +55,52 @@ public async Task<IEnumerable<Ilr>> Search(string searchQuery)
...
+            }
+            await _context.SaveChangesAsync();
+        }
+        private string createILRQuery(string filter)
+        {
+            return @"SELECT ilr.*
+                 FROM
+	                (
+		                SELECT *,
+		                ROW_NUMBER() OVER (PARTITION BY [Uln], [StdCode] ORDER BY [EventId] DESC, [Source] DESC, [CreatedAt] DESC, [LearnStartDate] DESC ) AS rownumber
```
Comment (martin-hornsby) [inline src/SFA.DAS.AssessorService.Data/IlrRepository.cs]: Am I right in thinking we no longer need DENSE_RANK() ?

File: src/SFA.DAS.AssessorService.Application/Handlers/Search/SearchHandler.cs [hunk 2]
```diff
@@ -97,8 +117,63 @@ await _ilrRepository.StoreSearchLog(new SearchLog()
...
+
+                listOfSearchResults.Add(result);
+            }
+
+            return listOfSearchResults;
+        }
+
         private async Task<List<int>> GetEpaoStandards(Organisation thisEpao)
         {
+            //Todo: get from our database don't need to go to external API
```
[the remark this answers is not in the document]
Comment (arshedmust) [reply src/SFA.DAS.AssessorService.Application/Handlers/Search/SearchHandler.cs]: Need to look at the criteria FindAllStandardsByOrganisationIdAsync uses to FindAll  and do a sanity check  against our locally stored data.

File: src/SFA.DAS.AssessorService.Application/Handlers/Search/SearchResultExtensions.cs [hunk 3]
```diff
@@ -20,83 +22,150 @@ public static List<SearchResult> PopulateStandards(this List<SearchResult> searc
...
             }
 
             return searchResults;
         }
 
-        public static List<SearchResult> MatchUpExistingCompletedStandards(this List<SearchResult> searchResults, SearchQuery request, ICertificateRepository certificateRepository, IContactQueryRepository contactRepository, ILogger<SearchHandler> logger)
+        public static List<SearchResult> CleanFromSearchResultAnyMarkedAsRemove(this List<SearchResult> searchResults)
+        {
+            for (var i = searchResults.Count - 1; i >= 0; i--)
+                if(searchResults[i].RemoveFromCollection)
```
[the remark this answers is not in the document]
Comment (arshedmust) [reply src/SFA.DAS.AssessorService.Application/Handlers/Search/SearchResultExtensions.cs]: I initially did use Linq but a bit of googling convinced me that a standard for loop is significantly faster and the recommended way

File: src/SFA.DAS.AssessorService.Data/CertificateRepository.cs [hunk 4]
```diff
@@ -147,7 +147,17 @@ private bool CheckCertificateData(Certificate certificate, string lastName, Date
...
+            return await _context.Certificates.Where(c =>
+                    c.Uln == uln && (c.Status == CertificateStatus.Reprint || c.Status == CertificateStatus.Printed ||
+                                     c.Status == CertificateStatus.Submitted))
+                .ToListAsync();
+        }
+
+        public async Task<List<Certificate>> GetCertificatesFor(long uln)
+        {
+            return await _context.Certificates.Where(c =>
+                    c.Uln == uln && (c.Status == CertificateStatus.Reprint || c.Status == CertificateStatus.Printed ||
```
[the remark this answers is not in the document]
Comment (arshedmust) [reply src/SFA.DAS.AssessorService.Data/CertificateRepository.cs]: Changed name to GetSubmittedAndDraftCertificatesFor

File: src/SFA.DAS.AssessorService.Data/IlrRepository.cs (hunk 1 above)
Comment (arshedmust) [reply src/SFA.DAS.AssessorService.Data/IlrRepository.cs]: Yes we don't need DENSE_RANK anymore.

File: src/SFA.DAS.AssessorService.Application/Interfaces/ICertificateRepository.cs [hunk 5]
```diff
@@ -31,5 +31,8 @@ public interface ICertificateRepository
         Task<CertificateAddress> GetContactPreviousAddress(string userName, bool requestIsPrivatelyFunded);
         Task<List<Option>> GetOptions(int stdCode);
         Task ApproveCertificates(List<ApprovalResult> approvalResults, string userName);
+        Task<List<Certificate>> GetCertificates(long uln, string familyName, int? standardCode = null);
```
[the remark this answers is not in the document]
Comment (arshedmust) [reply src/SFA.DAS.AssessorService.Application/Interfaces/ICertificateRepository.cs]: Looks as though it is not used anyway else, so removed

File: src/SFA.DAS.AssessorService.Data/CertificateRepository.cs [hunk 6]
```diff
@@ -421,5 +431,21 @@ private bool CheckLastNameExists(Certificate certificate, Certificate c)
             var certificateDataCompare = JsonConvert.DeserializeObject<CertificateData>(c.CertificateData);
             return certificateData.LearnerFamilyName == certificateDataCompare.LearnerFamilyName;
         }
+
+        public async Task<List<Certificate>> GetCertificates(long uln, string familyName, int? standardCode = null)
+        {
+            var certificate = standardCode  == null ? await
+                    _context.Certificates
```
[the remark this answers is not in the document]
Comment (arshedmust) [reply src/SFA.DAS.AssessorService.Data/CertificateRepository.cs]: Had a discussion

File: src/SFA.DAS.AssessorService.Settings/WebConfiguration.cs [hunk 7]
```diff
@@ -22,6 +22,8 @@ public class WebConfiguration : IWebConfiguration
 
         [JsonRequired] public string IfaApiClientBaseUrl { get; set; }
 
+        [JsonRequired] public string ProviderEventsClientBaseUrl { get; set; }
```
Comment (martin-hornsby) [inline src/SFA.DAS.AssessorService.Settings/WebConfiguration.cs]: Are you aware about the das-employer-config project?

I can walk you through it, however you'll need to update both the schema file and the json file in there then get one of us plus a person from DevOps to approve

**Comment by alandburns** [comment]
Closing this as I do not want it accidentally merging in.

**Pull request closed without merging**